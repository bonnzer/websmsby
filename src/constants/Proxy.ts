export const WEBSMS_GATEWAY = 'cp.websms.by';
export const WEBSMS_API_PATH = '/api';
