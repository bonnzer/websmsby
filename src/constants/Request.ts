export enum METHODS {
  POST = 'POST',
  GET = 'GET',
}

export enum BALANCE_TYPES {
  SMS = 0,
  VIBER,
}
