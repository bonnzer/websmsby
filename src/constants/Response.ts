export enum STATUSES {
  SUCCESS = 'success',
  ERROR = 'error',
}

export enum ERROR_CODES_COMMON {
  INCORRECT_LOGIN = 1,
  INCORRECT_APIKEY,
  SERVER_ERROR,
  VALIDATION_ERROR,
  OBJECT_NOT_FOUND,
  INCORRECT_REQUEST,
  MISSING_REQUIRED_PARAM,
}

export enum ERROR_CODES_CARDS_ACTIVATION {
  CARD_NOT_EXISTS = 16,
  CARD_ALREADY_ACTIVATED,
  CARD_ACTIVATED_BY_ANOTHER_USER,
}

export type ERROR_CODES = ERROR_CODES_COMMON & ERROR_CODES_CARDS_ACTIVATION;
