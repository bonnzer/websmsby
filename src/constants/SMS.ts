export enum SMS_STATUS {
  NEW = 'news',
  SEND = 'send',
  DELIVERED = 'delivered',
  NOTDELIVERED = 'notdelivered',
  BLOCKED = 'blocked',
  INPROGRESS = 'inprogress',
  ABSENT = 'absent',
}
