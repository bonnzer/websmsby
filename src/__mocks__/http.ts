import { METHODS } from '../constants/Request';

export default {
  request: jest.fn((args) => {
    expect(args).toMatchSnapshot();

    return {
      on: jest.fn((e, callback) => {
        callback('test');

        return {
          end: jest.fn(),
          write: jest.fn((body) => {
            if (e.method === METHODS.POST) {
              expect(body).toMatchSnapshot();
            }
          }),
        };
      }),
    };
  }),
};
