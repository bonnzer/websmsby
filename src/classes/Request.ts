import http from 'http';
import * as querystring from 'querystring';

import { WebsmsError } from './Error';
import { ErrorResponse, Response } from '../types/Response';
import { METHODS } from '../constants/Request';
import { Payload, RequestCredentials } from '../types/Request';

export class Request {
  private response = '';

  private options: http.RequestOptions;

  private data: string;

  constructor(
    options: http.RequestOptions,
    credentials: RequestCredentials,
    payload?: Payload,
  ) {
    const { method } = options;

    this.options = options;

    if (method === METHODS.GET) {
      const modifiedPayload: Record<string, unknown> = {};

      if (typeof payload !== 'undefined' && !Array.isArray(payload)) {
        const keys = Object.keys(payload);

        Object.values(payload).forEach((value, i) => {
          modifiedPayload[keys[i]] = Array.isArray(value)
            ? value.join(',')
            : value;
        });
      }

      this.options.path += `?${querystring.encode({
        ...credentials,
        ...modifiedPayload,
      })}`;

      this.data = '';
    } else {
      this.data = JSON.stringify({ ...credentials, ...payload });
      this.options.headers = {
        ...this.options.headers,
        'Content-length': this.data.length,
      };
    }
  }

  private handleData(d: string) {
    this.response += d;
  }

  public invoke(): Promise<Response> {
    return new Promise((resolve, reject) => {
      const req = http
        .request({ ...this.options }, (res) => {
          res.on('data', this.handleData).on('end', () => {
            try {
              const encoded: Response | ErrorResponse = JSON.parse(
                this.response,
              );

              return 'message' in encoded
                ? reject(new WebsmsError(encoded.message, encoded.error))
                : resolve(encoded);
            } catch (e) {
              return reject(e);
            }
          });
        })
        .on('error', (error) => reject(error));

      if (this.options.method === METHODS.POST) {
        req.write(this.data);
      }

      req.end();
    });
  }
}
