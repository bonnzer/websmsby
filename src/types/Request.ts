/* eslint-disable camelcase */
import { METHODS, BALANCE_TYPES } from '../constants/Request';

export type AccountActivationCardPayload = {
  cardno: string;
  balance_type: BALANCE_TYPES;
};

export type SendSMSPayload = {
  recipients: string | string[];
  message: string;
  sender: string;
  urgent?: 0 | 1;
  sendtime?: string;
};

export type GetSMSStatusPayload = {
  messages_id: string | string[];
};

export type GetSMSCostPayload = {
  recipients: string | string[];
  message: string;
};

export type GetSMSListPayload = {
  messages_id?: string | string[];
  bulk_id?: number;
  recipients?: string[];
  sender?: string | string[];
  status?: string;
  date_from?: string;
  date_to?: string;
  limit?: number;
  offset?: number;
  sort?: 'asc' | 'desc';
};

export type RequestCredentials = {
  user: string;
  apikey: string;
  test?: boolean;
  devkey?: string;
};

export type Payload =
  | AccountActivationCardPayload
  | SendSMSPayload
  | SendSMSPayload[]
  | GetSMSCostPayload
  | GetSMSListPayload
  | GetSMSStatusPayload
  | undefined;

export type SubmitParams = {
  method: METHODS;
  path: string;
  payload?: Payload;
};
