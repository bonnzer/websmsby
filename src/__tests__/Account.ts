import http from 'http';
import { Account } from '../index';
import { BALANCE_TYPES } from '../constants/Request';

jest.mock('http');

describe('Account', () => {
  describe('Params & body presence', () => {
    describe.each([
      [true, ''],
      [false, 'hey'],
    ])(
      `With testMode=%p and devKey=%p`,
      (testMode: boolean, devKey: string) => {
        const instance = new Account({
          user: 'test',
          apiKey: 'test',
          devKey,
          testMode,
        });

        const testMe = (method: () => void) => async () => {
          try {
            await method();
          } catch (e) {
            expect(e).toBe('test');
          }

          expect(http.request).toHaveBeenCalled();
        };

        test('getBalance()', testMe(instance.getBalance.bind(instance)));
        test('getUserNames()', testMe(instance.getUserNames.bind(instance)));
        test('getViberNames()', testMe(instance.getViberNames.bind(instance)));
        test(
          'activatePayCard()',
          testMe(
            instance.activatePayCard.bind(instance, {
              cardNo: 'testcard',
              balanceType: BALANCE_TYPES.SMS,
            }),
          ),
        );
      },
    );
  });
});
