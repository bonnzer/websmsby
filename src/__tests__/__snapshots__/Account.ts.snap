// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account Params & body presence With testMode=false and devKey="hey" activatePayCard() 1`] = `
Object {
  "headers": Object {
    "Content-type": "application/json",
  },
  "hostname": "cp.websms.by",
  "method": "GET",
  "path": "/api/paycard_activate?devkey=hey&apikey=test&test=false&user=test&cardno=testcard&balance_type=0",
}
`;

exports[`Account Params & body presence With testMode=false and devKey="hey" getBalance() 1`] = `
Object {
  "headers": Object {
    "Content-type": "application/json",
  },
  "hostname": "cp.websms.by",
  "method": "GET",
  "path": "/api/user_balance?devkey=hey&apikey=test&test=false&user=test",
}
`;

exports[`Account Params & body presence With testMode=false and devKey="hey" getUserNames() 1`] = `
Object {
  "headers": Object {
    "Content-type": "application/json",
  },
  "hostname": "cp.websms.by",
  "method": "GET",
  "path": "/api/user_names?devkey=hey&apikey=test&test=false&user=test",
}
`;

exports[`Account Params & body presence With testMode=false and devKey="hey" getViberNames() 1`] = `
Object {
  "headers": Object {
    "Content-type": "application/json",
  },
  "hostname": "cp.websms.by",
  "method": "GET",
  "path": "/api/viber_user_names?devkey=hey&apikey=test&test=false&user=test",
}
`;

exports[`Account Params & body presence With testMode=true and devKey="" activatePayCard() 1`] = `
Object {
  "headers": Object {
    "Content-type": "application/json",
  },
  "hostname": "cp.websms.by",
  "method": "GET",
  "path": "/api/paycard_activate?devkey=&apikey=test&test=true&user=test&cardno=testcard&balance_type=0",
}
`;

exports[`Account Params & body presence With testMode=true and devKey="" getBalance() 1`] = `
Object {
  "headers": Object {
    "Content-type": "application/json",
  },
  "hostname": "cp.websms.by",
  "method": "GET",
  "path": "/api/user_balance?devkey=&apikey=test&test=true&user=test",
}
`;

exports[`Account Params & body presence With testMode=true and devKey="" getUserNames() 1`] = `
Object {
  "headers": Object {
    "Content-type": "application/json",
  },
  "hostname": "cp.websms.by",
  "method": "GET",
  "path": "/api/user_names?devkey=&apikey=test&test=true&user=test",
}
`;

exports[`Account Params & body presence With testMode=true and devKey="" getViberNames() 1`] = `
Object {
  "headers": Object {
    "Content-type": "application/json",
  },
  "hostname": "cp.websms.by",
  "method": "GET",
  "path": "/api/viber_user_names?devkey=&apikey=test&test=true&user=test",
}
`;
